### PR TITLE
[RFC] Draft test of DB backup restore

### PR DESCRIPTION
**Pull request by lucafavatella**
In scope of https://www.pivotaltracker.com/story/show/153778514

----

This is the initial item for testing hard fork without restarting chain from genesis.

### CODE REVIEW

File: system_test/aest_fork_SUITE.erl [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    "ok" = aest_nodes:run_cmd_in_node_dir(old_node1, ["bin/epoch", "eval", "'{atomic, [_|_] = Tabs} = mnesia:restore(\"/tmp/mnesia_backup/mnesia_ae-uat-epoch_backup\", []), ok.'"], Cfg),
+    ct:log("~p", [aest_nodes:get_top(old_node1, Cfg)]), %% TODO Extract height at top.
+    _ = aest_nodes:get_block(old_node1, 100, Cfg), %% TODO Change to non-zero height at top.
+    ok.
+
+%=== INTERNAL FUNCTIONS ========================================================
+
+start_node(NodeName, Cfg) ->
+    aest_nodes:start_node(NodeName, Cfg),
+    F = fun() -> try aest_nodes:get_block(NodeName, 0, Cfg), true catch _:_ -> false end end,
```
Comment (sylane) [inline system_test/aest_fork_SUITE.erl]: You could use `aest_nodes:wait_for_height(0, [NodeName], SomeTimeout, Cfg)`, any reason not to ?

File: system_test/helpers/aest_docker.erl [hunk 2]
```diff
@@ -197,6 +200,21 @@ kill_node(#{container_id := ID, hostname := Name} = NodeState) ->
...
+    aest_docker_api:container_logs(ID).
+
+extract_archive(#{container_id := ID, hostname := Name} = NodeState, Path, Archive) ->
+    ok = aest_docker_api:extract_archive(ID, Path, Archive),
+    log(NodeState, "Extracted archive of size ~p in container ~p [~s] at path ~p", [byte_size(Archive), Name, ID, Path]),
+    NodeState.
+
+run_cmd_in_node_dir(#{container_id := ID, hostname := Name} = NodeState, Cmd) ->
+    log(NodeState, "Running command ~p on container ~p [~s]", [Cmd, Name, ID]),
+    Cmd1 = lists:flatten(io_lib:format("docker exec ~s ~s", [ID, lists:join($ , Cmd)])),
```
Comment (sylane) [inline system_test/helpers/aest_docker.erl]: I am adding a command exec using the HTTP API, the idea was to not use the command line API.

File: system_test/helpers/aest_nodes.erl [hunk 3]
```diff
@@ -148,6 +151,12 @@ stop_node(NodeName, Timeout, Ctx) ->
 kill_node(NodeName, Ctx) ->
     call(ctx2pid(Ctx), {kill_node, NodeName}).
 
+extract_archive(NodeName, Path, Archive, Ctx) ->
+    call(ctx2pid(Ctx), {extract_archive, NodeName, Path, Archive}).
+
+run_cmd_in_node_dir(NodeName, Cmd, Ctx) ->
```
Comment (sylane) [inline system_test/helpers/aest_nodes.erl]: I am adding an `exec` function that uses the docker API, is there any specific things this is doing I may be missing?

Comment (lucafavatella) [reply system_test/helpers/aest_nodes.erl]: stdout

File: system_test/aest_fork_SUITE.erl (hunk 1 above)
Comment (lucafavatella) [reply system_test/aest_fork_SUITE.erl]: The aehttp_integration_SUITE does same thing: interact meaningfully with user API of epoch node. Hence I was trying to reuse stuff from there.

File: Makefile [hunk 4]
```diff
@@ -151,9 +151,18 @@ eunit:
...
 	@./rebar3 as system_test do ct --dir system_test --logdir system_test/logs $(CT_TEST_FLAGS)
 
+system_test/aest_fork_SUITE_data/mnesia_ae-uat-epoch_backup.tar: system_test/aest_fork_SUITE_data/mnesia_ae-uat-epoch_backup
+	tar -c -C $(<D) -f $@ $(<F)
+
+system_test/aest_fork_SUITE_data/mnesia_ae-uat-epoch_backup: system_test/aest_fork_SUITE_data/mnesia_ae-uat-epoch_backup.gz
+	zcat <$< >$@
+
+system_test/aest_fork_SUITE_data/mnesia_ae-uat-epoch_backup.gz:
+	curl -fsS --create-dirs -o $@ https://5196-99802036-gh.circle-artifacts.com/0/tmp/chain_snapshots/ae-uat-epoch-n2/tmp/mnesia_ae-uat-epoch-n2_db_backup_1521936102.gz
```
[the remark this answers is not in the document]
Comment (dincho) [reply Makefile]: I don't think we have a fixed link of latest snapshot. I guess we have to handle this with some CircleCI api magic.

[the remark this answers is not in the document]
Comment (dincho) [reply Makefile]: e.g. https://circleci.com/docs/2.0/artifacts/#downloading-all-artifacts-for-a-build-on-circleci

**Comment by lucafavatella** [comment]
Thanks for the comments/feedback. I plan to open new PR with only docker logs section.

**Pull request closed without merging**